### PR TITLE
send ScalaMessagePack compact binary list of objects over the wire

### PR DESCRIPTION
**Pull request by eishay**
@ymatsuda @raymondng @yingjieMiao @stkem @andrewconner @leogrim 
Trying out ScalaMessagePack as a binary format.
See more at
http://wiki.msgpack.org/display/MSGPACK/QuickStart+for+Scala
http://msgpack.org/

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackIdTemplate.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        throw new MessageTypeException("Attempted to write null")
+      }
+      packer.writeNil()
+    }else{
+      packer.write(id.id)
+    }
+  }
+
+
+  def read(unpacker: Unpacker, to: Id[T], required: Boolean): Id[T] = {
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackIdTemplate.scala:17]: I see you're extending `AbstractTemplate`, so we may be limited, but any way we can avoid the `null`? Would be great to make this whole thing dealing with `Option`s.

Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackIdTemplate.scala:17]: Would it be okay to extend `AbstractTemplate[Option[Id[T]]`?

Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackIdTemplate.scala:17]: Extending `AbstractTemplate` is only for classes we don't want to or can't annotate with `@Message` as we do with `UriIdAndSeq`. For example, `DateTime` and other system classes. For regular message classes we'll use `@Message`.
As for nulls, you're absolutely right, we're not a Java shop here! I guess I was too tiered last night :-)
We should never have nulls and `required` must be always `true`. I'll have another `AbstractTemplate[Option[Id[T]]` for cases where value is optional.

**Comment by eishay** [comment]
@andrewconner please review diff

File: kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackSequenceNumberTemplate.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      packer.writeNil()
+    }else{
+      packer.write(seq.value)
+    }
+  }
+
+
+  def read(unpacker: Unpacker, to: SequenceNumber[T], required: Boolean): SequenceNumber[T] = {
+    if(!required && unpacker.trySkipNil){
+      return null
```
Comment (andrewconner) [inline kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackSequenceNumberTemplate.scala]: A bit cautious about introducing this null. We're accustomed to trusting variables as being non-null (or wrapped in options). Would prefer an exception here. Thoughts?

Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/model/serialize/MsgPackSequenceNumberTemplate.scala]: yes, `SequenceNumber` should be same as `Id` we should have no nulls. will fix

**Comment by eishay** [comment]
@andrewconner updated

**Comment by andrewconner** [comment]
Looks good.

**Comment by yingjieMiao** [comment]
lgtm

**Comment by eishay** [comment]
Merged,
@yingjieMiao please replace the call to `getScrapedFullURIs` with `getScrapedUriIdAndSeq`

**Comment by yingjieMiao** [comment]
A small problem: the URIPuller is committed to return a seq of 
NormalizedURIs. URIPuller works with featureUpdaters, featureStores, 
etc, which are typed with NormalizedURI. Direct use of that endpoint 
will give type mismatch.

A temp solution is to construct a dummy normalizedURi upon receive of 
UriIdAndSeq. So we don't have to change types all over the place.
next solution is to replace all the types.

**Pull request merged**